File: clients/elixir/lib/sharptown/response.ex
defmodule Sharptown.Response do
  @moduledoc """
  The result of a transform: the image bytes plus the response status and headers.
  Returned by `Sharptown.run/1` for every transport.
  """

  @enforce_keys [:status, :headers, :body]
  defstruct [:status, :headers, :body]

  @type t :: %__MODULE__{
          status: non_neg_integer(),
          headers: %{String.t() => String.t()},
          body: binary()
        }

  @doc "Whether the status is in the 2xx range."
  @spec ok?(t()) :: boolean()
  def ok?(%__MODULE__{status: status}), do: status in 200..299

  @doc "Fetches a response header by case-insensitive name."
  @spec header(t(), String.t()) :: String.t() | nil
  def header(%__MODULE__{headers: headers}, name), do: Map.get(headers, String.downcase(name))

  @doc "The `content-type` header, if present."
  @spec content_type(t()) :: String.t() | nil
  def content_type(%__MODULE__{} = response), do: header(response, "content-type")

  @doc "Writes the body to a file. Returns `:ok` or `{:error, reason}`."
  @spec to_file(t(), Path.t()) :: :ok | {:error, File.posix()}
  def to_file(%__MODULE__{body: body}, path), do: File.write(path, body)
end

File: clients/go/response.go
package sharptown

import (
	"net/http"
	"os"
)

// Response is the result of a transform: the image bytes plus the response status and
// headers. Returned by every transport.
type Response struct {
	Status int
	Header http.Header
	Body   []byte
}

// ContentType returns the response Content-Type header.
func (r *Response) ContentType() string { return r.Header.Get("Content-Type") }

// Save writes the body to a file.
func (r *Response) Save(path string) error { return os.WriteFile(path, r.Body, 0o644) }
